src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -374,85 +377,257 @@ internal static BigInteger ParseBigInteger(string value, NumberStyles style, Num
         }
 
-        internal static BigInteger ParseBigInteger(ReadOnlySpan<char> value, NumberStyles style, NumberFormatInfo info)
+        internal static (int sign, uint[]? bits) ParseBigInteger(ReadOnlySpan<char> value, NumberStyles style, NumberFormatInfo info)
         {
             ArgumentException? e;
             if (!TryValidateParseStyleInteger(style, out e))
                 throw e;
 
-            BigInteger result = BigInteger.Zero;
-            if (!TryParseBigInteger(value, style, info, out result))
+            if (!TryParseBigInteger(value, style, info, out int sign, out uint[]? bits))
             {
                 throw new FormatException(SR.Overflow_ParseBigInteger);
             }
-            return result;
+            return (sign, bits);
         }
 
-        private static unsafe bool HexNumberToBigInteger(ref BigNumberBuffer number, ref BigInteger value)
+        private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out int sign, out uint[]? bits)
         {
+            sign = 0;
+            bits = null;
+
             if (number.digits == null || number.digits.Length == 0)
                 return false;
 
-            int len = number.digits.Length - 1; // Ignore trailing '\0'
-            byte[] bits = new byte[(len / 2) + (len % 2)];
+            const int DigitsPerBlock = 8;
 
-            bool shift = false;
-            bool isNegative = false;
-            int bitIndex = 0;
+            int totalDigitCount = number.digits.Length - 1;   // Ignore trailing '\0'
+            int blockCount, partialDigitCount;
 
-            // Parse the string into a little-endian two's complement byte array
-            // string value     : O F E B 7 \0
-            // string index (i) : 0 1 2 3 4 5 <--
-            // byte[] (bitIndex): 2 1 1 0 0 <--
-            //
-            for (int i = len - 1; i > -1; i--)
+            if (totalDigitCount % DigitsPerBlock == 0)
+            {
+                blockCount = totalDigitCount / DigitsPerBlock;
+                partialDigitCount = 0;
+            }
+            else
             {
-                char c = number.digits[i];
-                int b = HexConverter.FromChar(c);
-                Debug.Assert(b != 0xFF);
-                if (i == 0 && (b & 0x08) == 0x08)
-                    isNegative = true;
+                blockCount = (totalDigitCount / DigitsPerBlock) + 1;
+                partialDigitCount = DigitsPerBlock - (totalDigitCount % DigitsPerBlock);
+            }
+
+            bool isNegative = HexConverter.FromChar(number.digits[0]) >= 8;
+            uint partialValue = (isNegative && partialDigitCount > 0) ? 0xFFFFFFFFu : 0;
+
+            int[]? arrayFromPool = null;
+
+            Span<int> bitsBuffer = (blockCount <= BigInteger.StackallocUInt32Limit)
+                ? stackalloc int[blockCount]
+                : (arrayFromPool = ArrayPool<int>.Shared.Rent(blockCount)).AsSpan(0, blockCount);
```

Comment:
Is this `int`, when everything else deals with `uint`, so we are more likely to get a pooled array?

Reply:
Yes. The use of `ArrayPool<int>` is much more common than `ArrayPool<uint>` (and the latter is probably nonexistent in the core libraries). Something similar is [used in RegexParser](https://github.com/dotnet/runtime/blob/f0679ab750cb91bcb0d1ee9c738cc0d5cd628da5/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L45-L48).